File: apps/server/sdk/go/webhooks.go
package fortyone

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	webhookSecretPrefix     = "whsec_"
	webhookSignatureVersion = "v1"
	webhookSecretBytes      = 32
	maxWebhookBodyBytes     = 256 * 1024
	defaultWebhookTolerance = 5 * time.Minute
)

type WebhookVerificationErrorCode string

const (
	WebhookInvalidSecret    WebhookVerificationErrorCode = "invalid_secret"
	WebhookInvalidHeaders   WebhookVerificationErrorCode = "invalid_headers"
	WebhookInvalidTimestamp WebhookVerificationErrorCode = "invalid_timestamp"
	WebhookStaleTimestamp   WebhookVerificationErrorCode = "stale_timestamp"
	WebhookInvalidSignature WebhookVerificationErrorCode = "invalid_signature"
	WebhookInvalidBody      WebhookVerificationErrorCode = "invalid_body"
)

// WebhookVerificationError exposes only a stable classification and safe
// summary. It never retains the signing secret, signature, or request body.
type WebhookVerificationError struct {
	Code WebhookVerificationErrorCode
	msg  string
}

func (err *WebhookVerificationError) Error() string { return err.msg }

// VerifiedWebhook identifies an authenticated delivery. Callers must still
// durably deduplicate ID before applying business effects.
type VerifiedWebhook struct {
	ID        uuid.UUID
	Timestamp time.Time
}

// WebhookVerifier verifies FortyOne's exact-byte HMAC envelope before JSON is
// decoded. It is safe for concurrent use.
type WebhookVerifier struct {
	secret    []byte
	tolerance time.Duration
	now       func() time.Time
}

func (*WebhookVerifier) String() string            { return "FortyOneWebhookVerifier{secret:[REDACTED]}" }
func (verifier *WebhookVerifier) GoString() string { return verifier.String() }

// NewWebhookVerifier creates a verifier for a show-once whsec_ signing secret.
func NewWebhookVerifier(secret string, tolerance time.Duration) (*WebhookVerifier, error) {
	decoded, err := decodeSigningSecret(secret)
	if err != nil {
		return nil, err
	}
	if tolerance == 0 {
		tolerance = defaultWebhookTolerance
	}
	if tolerance < time.Second || tolerance > time.Hour {
		return nil, &WebhookVerificationError{Code: WebhookInvalidTimestamp, msg: "webhook timestamp tolerance must be from one second through one hour"}
	}
	return &WebhookVerifier{secret: decoded, tolerance: tolerance, now: time.Now}, nil
}

// Verify validates Webhook-Id, Webhook-Timestamp, and every v1 value in
// Webhook-Signature. The raw body must be passed unchanged.
func (verifier *WebhookVerifier) Verify(body []byte, header http.Header) (VerifiedWebhook, error) {
	if len(body) == 0 || len(body) > maxWebhookBodyBytes {
		return VerifiedWebhook{}, &WebhookVerificationError{Code: WebhookInvalidBody, msg: "webhook body is empty or exceeds the supported limit"}
	}
	rawDeliveryID, ok := singleHeader(header, "Webhook-Id")
	if !ok || !isCanonicalUUID(rawDeliveryID) {
		return VerifiedWebhook{}, &WebhookVerificationError{Code: WebhookInvalidHeaders, msg: "Webhook-Id is missing or malformed"}
	}
	deliveryID, _ := uuid.Parse(rawDeliveryID)
	rawTimestamp, ok := singleHeader(header, "Webhook-Timestamp")
	if !ok {
		return VerifiedWebhook{}, &WebhookVerificationError{Code: WebhookInvalidTimestamp, msg: "Webhook-Timestamp is missing or malformed"}
	}
	timestampSeconds, err := parseWebhookTimestamp(rawTimestamp)
	if err != nil {
		return VerifiedWebhook{}, err
	}
	timestamp := time.Unix(timestampSeconds, 0).UTC()
	if absoluteDuration(verifier.now().Sub(timestamp)) > verifier.tolerance {
		return VerifiedWebhook{}, &WebhookVerificationError{Code: WebhookStaleTimestamp, msg: "webhook timestamp is outside the accepted replay window"}
	}

	mac := hmac.New(sha256.New, verifier.secret)
	_, _ = fmt.Fprintf(mac, "%s.%s.", rawDeliveryID, rawTimestamp)
	_, _ = mac.Write(body)
	expected := mac.Sum(nil)
	for _, headerValue := range header.Values("Webhook-Signature") {
		for _, candidate := range strings.Fields(headerValue) {
			parts := strings.Split(candidate, ",")
			if len(parts) != 2 || parts[0] != webhookSignatureVersion {
				continue
			}
			provided, decodeErr := base64.StdEncoding.Strict().DecodeString(parts[1])
			if decodeErr == nil && hmac.Equal(provided, expected) {
				return VerifiedWebhook{ID: deliveryID, Timestamp: timestamp}, nil
			}
		}
	}
	return VerifiedWebhook{}, &WebhookVerificationError{Code: WebhookInvalidSignature, msg: "webhook signature is invalid"}
}

func decodeSigningSecret(secret string) ([]byte, error) {
	if !strings.HasPrefix(secret, webhookSecretPrefix) {
		return nil, &WebhookVerificationError{Code: WebhookInvalidSecret, msg: "webhook signing secret is malformed"}
	}
	decoded, err := base64.StdEncoding.Strict().DecodeString(strings.TrimPrefix(secret, webhookSecretPrefix))
	if err != nil || len(decoded) != webhookSecretBytes {
		return nil, &WebhookVerificationError{Code: WebhookInvalidSecret, msg: "webhook signing secret is malformed"}
	}
	return decoded, nil
}

func parseWebhookTimestamp(raw string) (int64, error) {
	if raw == "" || len(raw) > 12 {
		return 0, &WebhookVerificationError{Code: WebhookInvalidTimestamp, msg: "Webhook-Timestamp is missing or malformed"}
	}
	for _, character := range raw {
		if character < '0' || character > '9' {
			return 0, &WebhookVerificationError{Code: WebhookInvalidTimestamp, msg: "Webhook-Timestamp is missing or malformed"}
		}
	}
	seconds, err := strconv.ParseInt(raw, 10, 64)
	if err != nil || seconds <= 0 {
		return 0, &WebhookVerificationError{Code: WebhookInvalidTimestamp, msg: "Webhook-Timestamp is missing or malformed"}
	}
	return seconds, nil
}

func singleHeader(header http.Header, name string) (string, bool) {
	values := header.Values(name)
	returnValue := ""
	if len(values) == 1 {
		returnValue = values[0]
	}
	return returnValue, len(values) == 1 && returnValue != "" && returnValue == strings.TrimSpace(returnValue)
}

func isCanonicalUUID(raw string) bool {
	if len(raw) != 36 || raw[8] != '-' || raw[13] != '-' || raw[18] != '-' || raw[23] != '-' {
		return false
	}
	parsed, err := uuid.Parse(raw)
	return err == nil && strings.EqualFold(parsed.String(), raw)
}

func absoluteDuration(value time.Duration) time.Duration {
	if value < 0 {
		return -value
	}
	return value
}

File: packages/sdk-typescript/src/webhooks.ts
const SIGNING_SECRET_PREFIX = "whsec_";
const SIGNATURE_VERSION = "v1";
const SIGNATURE_BYTES = 32;
const MAX_BODY_BYTES = 256 * 1024;
const DEFAULT_TOLERANCE_SECONDS = 5 * 60;
const UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/iu;

export type WebhookVerificationErrorCode =
  | "invalid_secret"
  | "invalid_headers"
  | "invalid_timestamp"
  | "stale_timestamp"
  | "invalid_signature"
  | "invalid_body";

export class WebhookVerificationError extends Error {
  readonly code: WebhookVerificationErrorCode;

  constructor(code: WebhookVerificationErrorCode, message: string) {
    super(message);
    this.name = "WebhookVerificationError";
    this.code = code;
  }
}

export interface VerifyWebhookInput {
  secret: string;
  body: Uint8Array;
  webhookId: string;
  webhookTimestamp: string;
  webhookSignature: string;
  now?: Date;
  toleranceSeconds?: number;
}

export interface VerifiedWebhook {
  id: string;
  timestamp: Date;
}

export const verifyWebhook = async (
  input: VerifyWebhookInput,
): Promise<VerifiedWebhook> => {
  if (!UUID_PATTERN.test(input.webhookId)) {
    throw new WebhookVerificationError(
      "invalid_headers",
      "Webhook-Id is missing or malformed",
    );
  }
  if (input.body.length === 0 || input.body.length > MAX_BODY_BYTES) {
    throw new WebhookVerificationError(
      "invalid_body",
      "Webhook body is empty or exceeds the supported limit",
    );
  }
  const timestampSeconds = parseTimestamp(input.webhookTimestamp);
  const timestamp = new Date(timestampSeconds * 1_000);
  const now = input.now ?? new Date();
  const toleranceSeconds =
    input.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
  if (
    !Number.isFinite(toleranceSeconds) ||
    toleranceSeconds < 1 ||
    toleranceSeconds > 3_600
  ) {
    throw new WebhookVerificationError(
      "invalid_timestamp",
      "Webhook timestamp tolerance must be from 1 through 3600 seconds",
    );
  }
  if (
    Math.abs(Math.floor(now.getTime() / 1_000) - timestampSeconds) >
    toleranceSeconds
  ) {
    throw new WebhookVerificationError(
      "stale_timestamp",
      "Webhook timestamp is outside the accepted replay window",
    );
  }

  const key = decodeSigningSecret(input.secret);
  const signedContent = concatenate(
    new TextEncoder().encode(
      `${input.webhookId}.${input.webhookTimestamp}.`,
    ),
    input.body,
  );
  let cryptoKey: CryptoKey;
  try {
    cryptoKey = await globalThis.crypto.subtle.importKey(
      "raw",
      key,
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["verify"],
    );
  } finally {
    key.fill(0);
  }
  for (const candidate of input.webhookSignature.trim().split(/\s+/u)) {
    const [version, encoded, extra] = candidate.split(",");
    if (version !== SIGNATURE_VERSION || !encoded || extra !== undefined) {
      continue;
    }
    const signature = decodeBase64(encoded);
    if (
      signature?.length === SIGNATURE_BYTES &&
      (await globalThis.crypto.subtle.verify(
        "HMAC",
        cryptoKey,
        signature,
        signedContent,
      ))
    ) {
      return { id: input.webhookId, timestamp };
    }
  }
  throw new WebhookVerificationError(
    "invalid_signature",
    "Webhook signature is invalid",
  );
};

const parseTimestamp = (value: string) => {
  if (!/^\d{1,12}$/u.test(value)) {
    throw new WebhookVerificationError(
      "invalid_timestamp",
      "Webhook-Timestamp is missing or malformed",
    );
  }
  const seconds = Number(value);
  if (!Number.isSafeInteger(seconds) || seconds <= 0) {
    throw new WebhookVerificationError(
      "invalid_timestamp",
      "Webhook-Timestamp is missing or malformed",
    );
  }
  return seconds;
};

const decodeSigningSecret = (secret: string) => {
  if (!secret.startsWith(SIGNING_SECRET_PREFIX)) {
    throw new WebhookVerificationError(
      "invalid_secret",
      "Webhook signing secret is malformed",
    );
  }
  const decoded = decodeBase64(secret.slice(SIGNING_SECRET_PREFIX.length));
  if (decoded?.length !== SIGNATURE_BYTES) {
    throw new WebhookVerificationError(
      "invalid_secret",
      "Webhook signing secret is malformed",
    );
  }
  return decoded;
};

const decodeBase64 = (value: string) => {
  if (
    value.length === 0 ||
    value.length % 4 !== 0 ||
    !/^[A-Za-z0-9+/]+={0,2}$/u.test(value)
  ) {
    return undefined;
  }
  try {
    const decoded = Uint8Array.from(globalThis.atob(value), (character) =>
      character.charCodeAt(0),
    );
    const canonical = globalThis.btoa(
      Array.from(decoded, (byte) => String.fromCharCode(byte)).join(""),
    );
    return canonical === value ? decoded : undefined;
  } catch {
    return undefined;
  }
};

const concatenate = (prefix: Uint8Array, body: Uint8Array) => {
  const result = new Uint8Array(prefix.length + body.length);
  result.set(prefix);
  result.set(body, prefix.length);
  return result;
};
